src/main/java/edu/umich/ctools/analytics/AnalyticsServlet.java

```diff
@@ -0,0 +1,329 @@
+			outputStream.flush();
+			outputStream.close();
+		} catch (IOException e) {
+			M_log.error("Writing the csv file to the has some errror: ",e);
+		}
+	}
+	protected void doPost(HttpServletRequest request,HttpServletResponse response) {
+		M_log.debug("doGet: Called");
+	}
+
+	private StringBuilder generateCSVFile(CoursesForSubaccounts cfs) {
+		M_log.debug("generateCSVFile: Called");
+		StringBuilder writer = new StringBuilder();
+		ArrayList<Course> courses = cfs.getCourse();
+		writer.append("COURSE_NAME");writer.append(',');
+		writer.append("COURSE_ID");writer.append(',');
+		writer.append("COURSE_URL");writer.append(',');
+		writer.append("INSTRUCTOR_NAME");writer.append(',');
+		writer.append("INSTRUCTOR_EMAIL");writer.append(',');
+		writer.append("ACCOUNT_NAME");writer.append(',');
+		writer.append("PARENT_ACCOUNT_NAME");
+		writer.append('\n');
+		int count=0;
+		for (Course course : courses) {
+			writer.append("\"" +course.getCourseName()+ "\"");writer.append(',');
+			writer.append(Integer.toString(course.getCourseId()));writer.append(',');
+			writer.append(course.getCourseUrl());writer.append(',');
+			writer.append(StringUtils.join(course.getInstructors().values(),';'));writer.append(',');
+			writer.append(StringUtils.join(course.getInstructors().keySet(),';'));writer.append(',');
+			writer.append("\"" +course.getAccountName()+"\"" );writer.append(',');
+			writer.append("\"" +course.getParentAccountName()+"\"" );writer.append(',');
+			writer.append('\n');
+			count++;
+		}
+		return writer;
+		
+	}
+	
+
+	private HttpResponse doApiCall(String url) {
+		HttpUriRequest clientRequest = null;
+		HttpResponse httpResponse=null;
+		try {
+			clientRequest = new HttpGet(url);
+			HttpClient client = new DefaultHttpClient();
+			final ArrayList<NameValuePair> nameValues = new ArrayList<NameValuePair>();
+			nameValues.add(new BasicNameValuePair("Authorization", "Bearer" + " " + canvasToken));
+			nameValues.add(new BasicNameValuePair("content-type", "application/json"));
+			for (final NameValuePair h : nameValues) {
+				clientRequest.addHeader(h.getName(), h.getValue());
+			}
+			try {
+				 httpResponse = client.execute(clientRequest);
+
+			} catch (IOException e) {
+				M_log.error("Canvas API call did not complete successfully", e);
+			}
+		} catch (Exception e) {
+			M_log.error("GET request has some exceptions", e);
+		}
+		return httpResponse;
+	}
+	private void getThePublishedCourseList(String url, CoursesForSubaccounts cfs) {
+		M_log.debug("getThePublishedCourseList: Called");
+		HttpResponse httpResponse = doApiCall(url);
+		ObjectMapper mapper = new ObjectMapper();
+		HttpEntity entity = httpResponse.getEntity();
+		List<HashMap<String, Object>> courseList=new ArrayList<HashMap<String, Object>>();
+		try {
+			String jsonResponseString = EntityUtils.toString(entity);
+			courseList = mapper.readValue(jsonResponseString,new TypeReference<List<Object>>(){});
+			for (HashMap<String, Object> course : courseList) {
```

Comment:
This is hard to read / test. Rules of thumb would be:
-  Never have if statements nested more than 2 deep.  Refactor nested code into another method or methods.
- Often it make a lot of sense to have a loop that, basically, just goes through a loop, sends individual objects to a method, and assembles the result.  Separating the looping from the details of processing makes understanding much easier.